packages/fxa-auth-server/scripts/move-customers-to-new-plan/move-customers-to-new-plan.ts

```diff
@@ -0,0 +1,273 @@
+import { StripeHelper } from '../../lib/payments/stripe';
+
+/**
+ * Firestore subscriptions contain additional expanded information
+ * on top of the base Stripe.Subscription type
+ */
+export interface FirestoreSubscription extends Stripe.Subscription {
+  customer: string;
+  plan: Stripe.Plan;
+  price: Stripe.Price;
+}
+
+export class CustomerPlanMover {
+  private config: ConfigType;
+  private firestore: Firestore;
+  private stripeQueue: PQueue;
+  private stripe: Stripe;
+
+  /**
+   * A converter to move customers from one plan to another
+   * @param sourcePlanId A Stripe plan or price ID which customers will be moved off of
+   * @param destinationPlanId A Stripe plan or price ID which customers will be moved to
+   * @param excludePlanIds A list of Stripe plan or price ID which if customers have will not be subscribed to the destination plan
+   * @param batchSize Number of subscriptions to fetch from Firestore at a time
+   * @param outputFile A CSV file to output a report of affected subscriptions to
+   * @param stripeHelper An instance of StripeHelper
+   * @param rateLimit A limit for number of stripe requests within the period of 1 second
+   * @param database A reference to the FXA database
+   */
+  constructor(
+    private sourcePlanId: string,
+    private destinationPlanId: string,
+    private excludePlanIds: string[],
+    private batchSize: number,
+    private outputFile: string,
+    private stripeHelper: StripeHelper,
+    private database: any,
+    rateLimit: number
+  ) {
+    this.stripe = stripeHelper.stripe;
+
+    const config = Container.get<ConfigType>(AppConfig);
+    this.config = config;
+
+    const firestore = Container.get<Firestore>(AuthFirestore);
+    this.firestore = firestore;
+
+    this.stripeQueue = new PQueue({
+      intervalCap: rateLimit,
+      interval: 1000, // Stripe measures it's rate limit per second
+    });
+  }
+
+  /**
+   * Move all customers from the source plan to the destination plan in batches
+   */
+  async convert() {
+    let startAfter: string | null = null;
+    let hasMore = true;
+
+    while (hasMore) {
+      const subscriptions = await this.fetchSubsBatch(startAfter);
+
+      startAfter = subscriptions.at(-1)?.id as string;
+      if (!startAfter) hasMore = false;
+
+      await Promise.all(
+        subscriptions.map((sub) => this.convertSubscription(sub))
+      );
+    }
+  }
+
+  /**
+   * Fetches subscriptions from Firestore paginated by batchSize
+   * @param startAfter ID of the last element of the previous batch for pagination
+   * @returns A list of subscriptions from firestore
+   */
+  async fetchSubsBatch(startAfter: string | null) {
+    const collectionPrefix = `${this.config.authFirestore.prefix}stripe-`;
+    const subscriptionCollection = `${collectionPrefix}subscriptions`;
+
+    const subscriptionSnap = await this.firestore
+      .collectionGroup(subscriptionCollection)
+      .where('plan.id', '==', this.sourcePlanId)
```

Comment:
We already have an index created for this?

Reply:
No, one would need to be created